shared/src/commonMain/kotlin/io/customer/shared/work/JobDispatcher.kt

```diff
@@ -0,0 +1,39 @@
+package io.customer.shared.work
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+
+/**
+ * Interface to make execution with coroutines easy. The class depends on [JobExecutor] to
+ * run in background and exposes simpler methods to run on background/main and handle exceptions in
+ * coroutines.
+ */
+internal interface JobDispatcher {
+    val jobExecutor: JobExecutor
+
+    /**
+     * Called whenever a coroutine fails with exception. Child classes can implement it to perform
+     * operations expected to be performed after coroutine has failed. However, this is to prevent
+     * unwanted crashes and behaviors. Child class should not rely on this and catch the exceptions
+     * if they are expecting the operations to fail in their coroutines.
+     */
+    fun onCoroutineFailed(exception: Throwable) {}
+}
```

Comment:
This may look weird in the first look, but I was trying something different to make switching easy. Looking at its usage might help understand this better.